Fix: Ignore "Math superscript (^) and subscript (_) characters are not allowed in text mode" error for citation key (#11948)

Satyamkumarnavneet opened the pull request:
This pull request addresses the issue [#11948](https://github.com/JabRef/jabref/issues/11948) of false positive integrity check errors for citation keys containing superscript (`^`) and subscript (`_`) characters in text mode. Specifically:

1. Updated `LatexIntegrityChecker#check` to:
   - Ignore the validation error `TTEM03` for `InternalField.KEY_FIELD`.
2. Added test cases to ensure:
   - Citation key fields containing `^` or `_` are ignored during validation.
   - Other fields (e.g., `TITLE`, `AUTHOR`, `JOURNAL`, `BOOKTITLE`, `NOTE`) are still checked for invalid LaTeX syntax.

### Mandatory checks

<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [x] I own the copyright of the code submitted and I licence it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if change is visible to the user)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [/] [Checked developer's documentation](https://devdocs.jabref.org/): Is the information available and up to date? If not, I outlined it in this pull request.
- [/] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request to the documentation repository.


<img width="1470" alt="Screenshot 2025-01-16 at 6 19 26 PM" src="https://github.com/user-attachments/assets/9c67f1e8-79c8-4d96-881d-4eea80b4a2b2" />

Satyamkumarnavneet commented on src/main/java/org/jabref/logic/integrity/LatexIntegrityChecker.java:
```diff
@@ -62,6 +63,8 @@ public class LatexIntegrityChecker implements EntryChecker {
     @Override
     public List<IntegrityMessage> check(BibEntry entry) {
         return entry.getFieldMap().entrySet().stream()
+                    // Ignore checks for the citation key field
+                    .filter(field -> !(field.getKey().equals(InternalField.KEY_FIELD) && field.getValue().matches(".*[\\^_].*")))
```
Hi @koppor, Thanks for the feedback! Could you share any suggestions on how to better address the TTEM03 handling?

Satyamkumarnavneet replied:
Hi @koppor, I am only able to achieve one thing at a time; either it’s all the tests with InternalField and StandardField or the results shown in the screenshots.

Satyamkumarnavneet commented on src/test/java/org/jabref/logic/integrity/LatexIntegrityCheckerTest.java:
```diff
@@ -300,10 +300,18 @@ private static Stream<Arguments> provideUnacceptedInputs() {
                 // TFETB0: \hline must be the only token in table row
                 // Skipped
+
+                Arguments.of(LatexIntegrityChecker.errorMessageFormatHelper(CoreErrorCode.TTEM03), StandardField.TITLE, "Title with ^ and _ characters"),
+                // Ensure invalid LaTeX syntax with special characters in other fields is detected
+                Arguments.of(LatexIntegrityChecker.errorMessageFormatHelper(CoreErrorCode.TTEM03), StandardField.AUTHOR, "Author_Name"),
+                Arguments.of(LatexIntegrityChecker.errorMessageFormatHelper(CoreErrorCode.TTEM03), StandardField.JOURNAL, "Journal_Name"),
+                Arguments.of(LatexIntegrityChecker.errorMessageFormatHelper(CoreErrorCode.TTEM03), StandardField.BOOKTITLE, "Book^Title"),
+                Arguments.of(LatexIntegrityChecker.errorMessageFormatHelper(CoreErrorCode.TTEM03), StandardField.NOTE, "Note_with_subscript_and_superscript^_")
```
Thanks for pointing that out! I've added a new test case for InternalField.KEY_FIELD, but it's currently failing. Any suggestions on how to resolve this?

koppor replied:
I assume, it is not failing now?

Satyamkumarnavneet commented on src/test/java/org/jabref/logic/integrity/LatexIntegrityCheckerTest.java:
```diff
@@ -300,10 +300,18 @@ private static Stream<Arguments> provideUnacceptedInputs() {
+                Arguments.of(LatexIntegrityChecker.errorMessageFormatHelper(CoreErrorCode.TTEM03), StandardField.JOURNAL, "Journal_Name"),
+                Arguments.of(LatexIntegrityChecker.errorMessageFormatHelper(CoreErrorCode.TTEM03), StandardField.BOOKTITLE, "Book^Title"),
+                Arguments.of(LatexIntegrityChecker.errorMessageFormatHelper(CoreErrorCode.TTEM03), StandardField.NOTE, "Note_with_subscript_and_superscript^_")
+
```
Thank you for the feedback! I’ll make sure to take care of this next time.

Satyamkumarnavneet commented:
Hi @koppor, I have addressed all feedback and made the necessary changes. Could you please review the updates at your convenience? Thank you!

koppor commented on src/main/java/org/jabref/logic/integrity/LatexIntegrityChecker.java:
```diff
@@ -66,6 +65,9 @@ public List<IntegrityMessage> check(BibEntry entry) {
                     .filter(pair -> !pair.getValue().getErrorCode().getErrorGroup().equals(CoreErrorGroup.TDE))
+                    // Exclude TTEM03 error for citation key field
+                    .filter(pair -> !(pair.getValue().getErrorCode().equals(CoreErrorCode.TTEM03) && pair.getKey().equals(InternalField.KEY_FIELD)))
+                    // Exclude errors present in EXCLUDED_ERRORS set
```
Why this comment? This is really what is in the next line - normally one adds the **why** on top - or summarizes the next line. Please just remove line 70

Satyamkumarnavneet replied:
Got it! Thanks for pointing that out! I’ll remove line 70.

koppor commented on src/test/java/org/jabref/logic/integrity/LatexIntegrityCheckerTest.java:
```diff
@@ -300,10 +301,33 @@ private static Stream<Arguments> provideUnacceptedInputs() {
+    @MethodSource("provideCitationKeyInputs")
+    void citationKeyField(String expectedMessage, Field field, String value) {
+        entry.setField(field, value);
+        // Ensure TTEM03 error is not raised for citation key field
```
This comment has to move above `Arguments.of("", InternalField.KEY_FIELD, "Corti_2009"),` to explain the `_` in the inputs. -- In this place, it does not make sense.

Satyamkumarnavneet replied:
Yes, it's not failing now. It's working for all pre-existing tests as well as the newly added ones.

<img width="983" alt="Screenshot 2025-02-03 at 7 34 58 PM" src="https://github.com/user-attachments/assets/2d53561c-46ab-4b31-b6a1-1b1fb90de72f" />

Satyamkumarnavneet commented:
Hi @koppor, I've made the requested changes: removed the comment in `LatexIntegrityChecker.java`, moved the test comment, and confirmed TTEM03 no longer triggers for citation keys. All tests are passing. Please proceed with the merge. Thanks!
